# Task 6.1: Spotify Analysis- Sourcing Open Data

In [1]:
#Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create a shortcut for project folder path as a string.
path = r'/Users/cboyd/Documents/05-2022 Spotify Analysis'

## Import data

In [4]:
# Import the “Database to calculate popularity.csv” data set into your notebook using the os library.
df_pop = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Database to calculate popularity.csv'), index_col = False)

In [5]:
# Test DF
df_pop

Unnamed: 0 country        date  position  \
0                 0  Global  05/11/2020       1.0   
1                 1  Global  05/11/2020       2.0   
2                 2  Global  05/11/2020       3.0   
3                 3  Global  05/11/2020       4.0   
4                 4  Global  05/11/2020       5.0   
...             ...     ...         ...       ...   
9806996     9806996  Taiwan  01/01/2017     197.0   
9806997     9806997  Taiwan  01/01/2017     198.0   
9806998     9806998  Taiwan  01/01/2017     199.0   
9806999     9806999  Taiwan  01/01/2017     200.0   
9807000     9807000     NaN         NaN       NaN   

                                                       uri  \
0        https://open.spotify.com/track/47EiUVwUp4C9fGc...   
1        https://open.spotify.com/track/35mvY5S1H3J2QZy...   
2        https://open.spotify.com/track/3tjFYV6RSFtuktY...   
3        https://open.spotify.com/track/7hxHWCCAIIxFLCz...   
4        https://open.spotify.com/track/4y4spB9m0Q6026K...   
...                                                    ...   
9806996  https://open.spotify.com/track/3zHq9ouUJQFQRf3...   
9806997  https://open.spotify.com/track/3CRDbSIZ4r5MsZ0...   
9806998  https://open.spotify.com/track/2mEvsc8k8iwR099...   
9806999  https://open.spotify.com/track/0L7loeTIg6akW84...   
9807000                                                NaN   

                                                     track  \
0                         Dakiti , Bad Bunny, Jhay Cortez    
1                               positions , Ariana Grande    
2                                        Mood  , 24kGoldn    
3                              Lemonade  , Internet Money    
4                                 Lonely  , Justin Bieber    
...                                                    ...   
9806996   Love Me Like You Do - From "Fifty Shades Of G...   
9806997                  Stressed Out , Twenty One Pilots    
9806998   The End - From "La La Land" Soundtrack , Just...   
9806999                Light It Up  - Remix , Major Lazer    
9807000                                                NaN   

                                                     title  \
0                                                  Dakiti    
1                                               positions    
2                                                   Mood     
3                                               Lemonade     
4                                                 Lonely     
...                                                    ...   
9806996   Love Me Like You Do - From "Fifty Shades Of G...   
9806997                                      Stressed Out    
9806998            The End - From "La La Land" Soundtrack    
9806999                              Light It Up  - Remix    
9807000                                                NaN   

                           artist  
0         Bad Bunny, Jhay Cortez   
1                  Ariana Grande   
2                       24kGoldn   
3                 Internet Money   
4                  Justin Bieber   
...                           ...  
9806996           Ellie Goulding   
9806997        Twenty One Pilots   
9806998           Justin Hurwitz   
9806999              Major Lazer   
9807000                       NaN  

[9807001 rows x 8 columns]

In [6]:
# Import the “Final database.csv” data set into your notebook using the os library.
df_final = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'Final database.csv'), index_col = False)

/Users/cboyd/opt/anaconda3/lib/python3.9/site-packages/IPython/core/interactiveshell.py:3444: DtypeWarning: Columns (7,8,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


In [7]:
# Test DF
df_final

Country                                                Uri  \
0            Global  https://open.spotify.com/track/6FyRXC8tJUh863J...   
1               USA  https://open.spotify.com/track/6FyRXC8tJUh863J...   
2         Argentina  https://open.spotify.com/track/6FyRXC8tJUh863J...   
3           Belgium  https://open.spotify.com/track/6FyRXC8tJUh863J...   
4       Switzerland  https://open.spotify.com/track/6FyRXC8tJUh863J...   
...             ...                                                ...   
170628       Taiwan  https://open.spotify.com/track/4v29A51TqtD2KFS...   
170629       Taiwan  https://open.spotify.com/track/7InvbALrXpsz58J...   
170630       Taiwan  https://open.spotify.com/track/5rCNIuDc3VD7q6H...   
170631       Taiwan  https://open.spotify.com/track/0525Cll5exsrybt...   
170632       Taiwan  https://open.spotify.com/track/6IUFuke9UsBg2FD...   

        Popularity       Title            Artist Album/Single  \
0         31833.95  adan y eva      Paulo Londra       single   
1             8.00  adan y eva      Paulo Londra       single   
2         76924.40  adan y eva      Paulo Londra       single   
3           849.60  adan y eva      Paulo Londra       single   
4         20739.10  adan y eva      Paulo Londra       single   
...            ...         ...               ...          ...   
170628       31.20   smalltown  Luke Christopher       single   
170629      156.00          灰狼            G.E.M.        album   
170630       90.40        預念患者               張赫宣        album   
170631       21.60        清晨巴士          Crowd Lu        album   
170632       46.40   gettin by               ONE       single   

                    Genre Artist_followers Explicit       Album  ... Sweden  \
0       argentine hip hop       11427104.0    False  Adan y Eva  ...      0   
1       argentine hip hop       11427104.0    False  Adan y Eva  ...      0   
2       argentine hip hop       11427104.0    False  Adan y Eva  ...      0   
3       argentine hip hop       11427104.0    False  Adan y Eva  ...      0   
4       argentine hip hop       11427104.0    False  Adan y Eva  ...      0   
...                   ...              ...      ...         ...  ...    ...   
170628      indie pop rap         152740.0     True   SmallTown  ...      0   
170629           mandopop        1225408.0    False       摩天動物園  ...      0   
170630                n-a             38.0    False     Freedom  ...      0   
170631      chinese indie         149875.0    False    有吉他的流行歌曲  ...      0   
170632              k-rap         205012.0    False     ONE DAY  ...      0   

       Switzerland Taiwan Turkey UK USA Popu_max Top10_dummy Top50_dummy  \
0                0      0      0  0   0       10         1.0         1.0   
1                0      0      0  0   1      191         0.0         0.0   
2                0      0      0  0   0        1         1.0         1.0   
3                0      0      0  0   0      126         0.0         0.0   
4                1      0      0  0   0       21         0.0         1.0   
...            ...    ...    ... ..  ..      ...         ...         ...   
170628           0      1      0  0   0      162         0.0         0.0   
170629           0      1      0  0   0      130         0.0         0.0   
170630           0      1      0  0   0      133         0.0         0.0   
170631           0      1      0  0   0      174         0.0         0.0   
170632           0      1      0  0   0      143         0.0         0.0   

                            Cluster  
0                            global  
1       english speaking and nordic  
2                  spanish speaking  
3       english speaking and nordic  
4       english speaking and nordic  
...                             ...  
170628  english speaking and nordic  
170629  english speaking and nordic  
170630  english speaking and nordic  
170631  english speaking and nordic  
170632  english speaking and nordic  

[170633 rows x 151 colu

## Data profiling

In [8]:
# Check df_pop columns
df_pop.columns

Index(['Unnamed: 0', 'country', 'date', 'position', 'uri', 'track', 'title',
       'artist'],
      dtype='object')

In [9]:
# Check df_final columns
df_final.columns

Index(['Country', 'Uri', 'Popularity', 'Title', 'Artist', 'Album/Single',
       'Genre', 'Artist_followers', 'Explicit', 'Album',
       ...
       'Sweden', 'Switzerland', 'Taiwan', 'Turkey', 'UK', 'USA', 'Popu_max',
       'Top10_dummy', 'Top50_dummy', 'Cluster'],
      dtype='object', length=151)

In [11]:
# Aggregates for df_pop 
df_pop.describe()

Unnamed: 0      position
count  9.807001e+06  9.807000e+06
mean   4.903500e+06  1.005000e+02
std    2.831037e+06  5.773431e+01
min    0.000000e+00  1.000000e+00
25%    2.451750e+06  5.075000e+01
50%    4.903500e+06  1.005000e+02
75%    7.355250e+06  1.502500e+02
max    9.807000e+06  2.000000e+02

In [13]:
# Aggregates for df_final
df_final.describe()

Popularity  Days_since_release  Released_after_2017  Explicit_false  \
count  170633.000000       167411.000000        167411.000000   170633.000000   
mean     5417.616264         1337.530228             0.845667        0.651287   
std     13115.854526         2453.554101             0.361269        0.476564   
min         0.800000            9.000000             0.000000        0.000000   
25%        77.600000          428.000000             1.000000        0.000000   
50%       515.150000          834.000000             1.000000        1.000000   
75%      3867.850000         1240.000000             1.000000        1.000000   
max    233766.900000        44128.000000             1.000000        1.000000   

       Explicit_true          album    compilation         single  \
count  170633.000000  170633.000000  170633.000000  170633.000000   
mean        0.348684       0.549149       0.015026       0.435795   
std         0.476555       0.497580       0.121658       0.495862   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       1.000000       0.000000       0.000000   
75%         1.000000       1.000000       0.000000       1.000000   
max         1.000000       1.000000       1.000000       1.000000   

              bolero       boy band  ...         Spain         Sweden  \
count  170633.000000  170633.000000  ...  170633.00000  170633.000000   
mean        0.001242       0.011621  ...       0.02513       0.040848   
std         0.035226       0.107175  ...       0.15652       0.197938   
min         0.000000       0.000000  ...       0.00000       0.000000   
25%         0.000000       0.000000  ...       0.00000       0.000000   
50%         0.000000       0.000000  ...       0.00000       0.000000   
75%         0.000000       0.000000  ...       0.00000       0.000000   
max         1.000000       1.000000  ...       1.00000       1.000000   

         Switzerland         Taiwan         Turkey             UK  \
count  170633.000000  170633.000000  170633.000000  170633.000000   
mean        0.045044       0.044505       0.019721       0.038328   
std         0.207401       0.206215       0.139039       0.191987   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

                 USA       Popu_max    Top10_dummy    Top50_dummy  
count  170633.000000  170633.000000  170633.000000  170633.000000  
mean        0.038943      89.926738       0.087035       0.294843  
std         0.193460      57.998729       0.281887       0.455974  
min         0.000000       1.000000       0.000000       0.000000  
25%         0.000000      40.000000       0.000000       0.000000  
50%         0.000000      85.000000       0.000000       0.000000  
75%         0.000000     138.000000       0.000000       1.000000  
max         1.000000     200.000000       1.000000       1.000000  

[8 rows x 123 columns]

## Data Cleaning

In [14]:
# Check for missing values df_pop
df_pop.isnull().sum()

Unnamed: 0        0
country           1
date              1
position          1
uri               1
track             1
title             1
artist        15642
dtype: int64

In [21]:
# Create a subset of missing values in df_pop
df_pop_nan = df_pop[df_pop['artist'].isnull() == True]

In [22]:
# Check above work
df_pop_nan

Unnamed: 0 country        date  position  uri track title artist
176000       176000  Global  09/06/2018       1.0  N\A   N\A   N\A    NaN
176001       176001  Global  09/06/2018       2.0  N\A   N\A   N\A    NaN
176002       176002  Global  09/06/2018       3.0  N\A   N\A   N\A    NaN
176003       176003  Global  09/06/2018       4.0  N\A   N\A   N\A    NaN
176004       176004  Global  09/06/2018       5.0  N\A   N\A   N\A    NaN
...             ...     ...         ...       ...  ...   ...   ...    ...
9790996     9790996  Taiwan  23/03/2017     197.0  N\A   N\A   N\A    NaN
9790997     9790997  Taiwan  23/03/2017     198.0  N\A   N\A   N\A    NaN
9790998     9790998  Taiwan  23/03/2017     199.0  N\A   N\A   N\A    NaN
9790999     9790999  Taiwan  23/03/2017     200.0  N\A   N\A   N\A    NaN
9807000     9807000     NaN         NaN       NaN  NaN   NaN   NaN    NaN

[15642 rows x 8 columns]

In [17]:
# Check for duplicates df_pop
df_pop_dups = df_pop[df_pop.duplicated()]

In [18]:
# Test new df, look for dups
df_pop_dups

Empty DataFrame
Columns: [Unnamed: 0, country, date, position, uri, track, title, artist]
Index: []

### There is 1 missing value for 6 variables and 15,642 for the "artist" variable. There are no duplicates present in this DF. 

In [20]:
# Check data types for df_pop
df_pop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9807001 entries, 0 to 9807000
Data columns (total 8 columns):
 #   Column      Dtype  
---  ------      -----  
 0   Unnamed: 0  int64  
 1   country     object 
 2   date        object 
 3   position    float64
 4   uri         object 
 5   track       object 
 6   title       object 
 7   artist      object 
dtypes: float64(1), int64(1), object(6)
memory usage: 598.6+ MB


### This dataframe looks pretty clean already and shouldn't be an issue with the current data types.

In [23]:
# Check for missing values df_final
df_final.isnull().sum()

Country        0
Uri            0
Popularity     0
Title          0
Artist         0
              ..
USA            0
Popu_max       0
Top10_dummy    0
Top50_dummy    0
Cluster        0
Length: 151, dtype: int64

### There are no missing values 

In [25]:
# Check for duplicates df_final
df_final_dups = df_final[df_final.duplicated()]

In [26]:
# Test new df, look for dups
df_final_dups

Empty DataFrame
Columns: [Country, Uri, Popularity, Title, Artist, Album/Single, Genre, Artist_followers, Explicit, Album, Release_date, Track_number, Tracks_in_album, danceability, energy, key, loudness, mode, speechiness, acoustics, instrumentalness, liveliness, valence, tempo, duration_ms, time_signature, Genre_new, Days_since_release, Released_after_2017, Explicit_false, Explicit_true, album, compilation, single, bolero, boy band, country, dance/electronic, else, funk, hip hop, house, indie, jazz, k-pop, latin, metal, opm, pop, r&b/soul, rap, reggae, reggaeton, rock, trap, syuzhet_norm, bing_norm, afinn_norm, nrc_norm, syuzhet, bing, afinn, nrc, anger, anticipation, disgust, fear, joy, sadness, surprise, trust, negative, positive, n_words, anger_norm, anticipation_norm, disgust_norm, fear_norm, joy_norm, sadness_norm, surprise_norm, trust_norm, negative_norm, positive_norm, anger_norm2, anticipation_norm2, disgust_norm2, fear_norm2, joy_norm2, sadness_norm2, surprise_norm2, trust_norm2, negative_norm2, positive_norm2, negative_bog_jr, positive_bog_jr, Bayes, Negative_Bayes, Neutral_Bayes, Positive_Bayes, ...]
Index: []

[0 rows x 151 columns]

### There are no duplicates present in the DF

In [27]:
# Check data types for df_final
df_final.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 170633 entries, 0 to 170632
Columns: 151 entries, Country to Cluster
dtypes: float64(58), int64(65), object(28)
memory usage: 196.6+ MB


In [29]:
for col in df_final.columns.tolist():
  weird = (df_final[[col]].applymap(type) != df_final[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_final[weird]) > 0:
    print (col)

Artist_followers
Explicit
Track_number
Tracks_in_album
danceability
energy
key
loudness
mode
speechiness
acoustics
instrumentalness
liveliness
valence
tempo
duration_ms
time_signature
LDA_Topic


In [30]:
# Re-check data types
df_final.dtypes

Country         object
Uri             object
Popularity     float64
Title           object
Artist          object
                ...   
USA              int64
Popu_max         int64
Top10_dummy    float64
Top50_dummy    float64
Cluster         object
Length: 151, dtype: object

In [31]:
# Adjust data type
df_final['Artist_followers'] = df_final['Artist_followers'].astype('str')

In [32]:
# Adjust data type
df_final['Explicit'] = df_final['Explicit'].astype('str')

In [33]:
# Adjust data type
df_final['LDA_Topic'] = df_final['LDA_Topic'].astype('str')

## Save work 

In [34]:
# Save df_pop as pickle
df_pop.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'Database to calculate popularity_clean.pkl'))

In [35]:
# Save df_final as pickle
df_final.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'Final database_clean.pkl'))